In [46]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Manual Assessment(Just to understand your data)

In [3]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

# Quality

## Patients
- Zip code columns is stored in decimal format and certain zipcode are of four digits
- For State columns sometimes full name is used sometimes short form is used.
- Missing contact information of 12 index
- Incorrect data type for assigned_sex as it is categorical and it is also incorrect fo birthdate
- Errorness in data for weight and height
- Inconsistent way of stroiong phone number in conatct column
- Duplicate and non-useful data in dataset

## Treatments
 - hba1c_change has missing values
 -  auralin	novodra columns has u attached to numbers
 - Only 280 rows are there.
 - Name and Surname are in lowercase...................
 - auralin	novodra incoorect datatype
 - hb1a 9 instead of 4
 - Use of dashes instead of NAN 

## AdverseEffects

# Programmatic Assessment

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [10]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [11]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [12]:
patients.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
498    False
499    False
500    False
501    False
502    False
Length: 503, dtype: bool

In [13]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

# Tidyness (Structural Issues)

## Patients
- Two variables are in single columns(phone and email)

## Treatments
- Three variables stored in Aurinlin and NOvodra columns

## Adverse_Reaction
- No point of using this Table

# CLEANING THE DATA

## Completeness Issues
- Missing contact details of 12 patients ->This is not solvable
- only 280 rows are there 
- hba1c col has missing values

In [14]:
patients_copy = pd.read_csv('patients.csv').copy()
treatments_copy = pd.read_csv('treatments.csv').copy()
adverse_reactions_copy = pd.read_csv('adverse_reactions.csv').copy()

In [15]:
treatments_clean_concat = None
treatments_clean_concat = pd.concat([treatments, pd.read_csv('treatments_cut.csv')], ignore_index=True)
treatments_clean = treatments_clean_concat

In [16]:
treatments_clean

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika    jindrová  41u - 48u          -         7.63       7.20   
1       elliot  richardson          -  40u - 45u         7.56       7.09   
2     yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3         skye  gormanston  33u - 36u          -         7.97       7.62   
4       alissa      montez          -  33u - 29u         7.78       7.46   
..         ...         ...        ...        ...          ...        ...   
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN  

[350 rows x 7 columns]

In [17]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

In [18]:
treatments_clean.describe()

hba1c_start   hba1c_end  hba1c_change
count   350.000000  350.000000    350.000000
mean      7.956343    7.560057      0.396286
std       0.545328    0.545456      0.059283
min       7.500000    7.010000      0.200000
25%       7.650000    7.270000      0.350000
50%       7.785000    7.400000      0.390000
75%       7.950000    7.557500      0.440000
max       9.950000    9.580000      0.540000

In [19]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


## Tidiness Issues
- Phone and email in same col
- Novodra and Aoro not correct
- Merge the adverse effect to treatment table

In [20]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], 
                           var_name='treatment', value_name='dose')

In [21]:
treatments_clean = treatments_clean[treatments_clean['dose'] != "-"]

In [22]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
3           skye  gormanston         7.97       7.62          0.35   auralin   
6         sophia      haugen         7.65       7.27          0.38   auralin   
7          eddie      archer         7.89       7.55          0.34   auralin   
9           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
688  christopher    woodward         7.51       7.06          0.45   novodra   
690        maret    sultygov         7.67       7.30          0.37   novodra   
694        lixue       hsueh         9.21       8.80          0.41   novodra   
696        jakob    jakobsen         7.96       7.51          0.45   novodra   
698        berta  napolitani         7.68       7.21          0.47   novodra   

          dose  
0    41u - 48u  
3    33u - 36u  
6    37u - 42u  
7    31u - 38u  
9    30u - 36u  
..         ...  
688  55u - 51u  
690  26u - 23u  
694  22u - 23u  
696  28u - 26u  
698  42u - 44u  

[350 rows x 7 columns]

In [23]:
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split('-').str

C:\Users\user\AppData\Local\Temp\ipykernel_16440\2239936580.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split('-').str


In [24]:
treatments_clean.drop('dose', axis=1,inplace = True)

In [25]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
3           skye  gormanston         7.97       7.62          0.35   auralin   
6         sophia      haugen         7.65       7.27          0.38   auralin   
7          eddie      archer         7.89       7.55          0.34   auralin   
9           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
688  christopher    woodward         7.51       7.06          0.45   novodra   
690        maret    sultygov         7.67       7.30          0.37   novodra   
694        lixue       hsueh         9.21       8.80          0.41   novodra   
696        jakob    jakobsen         7.96       7.51          0.45   novodra   
698        berta  napolitani         7.68       7.21          0.47   novodra   

    dose_start dose_end  
0         41u       48u  
3         33u       36u  
6         37u       42u  
7         31u       38u  
9         30u       36u  
..         ...      ...  
688       55u       51u  
690       26u       23u  
694       22u       23u  
696       28u       26u  
698       42u       44u  

[350 rows x 8 columns]

In [26]:
patients_copy['phone_number'] = patients_copy.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_copy['email'] = patients_copy.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_copy = patients_copy.drop('contact', axis=1)

In [27]:
patients_copy

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country  birthdate  weight  height   bmi       phone_number  \
0    United States  7/10/1976   121.7      66  19.6       951-719-9170   
1    United States   4/3/1967   118.8      66  19.2  +1 (217) 569-3204   
2    United States  2/19/1980   177.8      71  24.8       402-363-6804   
3    United States  7/26/1951   220.9      70  31.7  +1 (732) 636-8246   
4    United States  2/18/1928   192.3      27  26.1       334-515-7487   
..             ...        ...     ...     ...   ...                ...   
498  United States  4/10/1959   181.1      72  24.6       207-477-0579   
499  United States  3/26/1948   239.6      70  34.4       928-284-4492   
500  United States  1/13/1971   171.2      67  26.8       816-223-6007   
501  United States  2/13/1952   176.9      67  27.7       360 443 2060   
502  United States   5/3/1954   138.2      71  19.3       402-848-4923   

                                 email  
0             ZoeWellish@superrito.com  
1                 PamelaSHill@cuvox.de  
2                 JaeMDebord@gustr.com  
3            PhanBaLiem@jourrapide.com  
4                  TimNeudorf@cuvox.de  
..                                 ...  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com  

[503 rows x 15 columns]

In [28]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_copy, 
                            on=['given_name', 'surname'], how='left')

In [29]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  christopher    woodward         7.51       7.06          0.45   novodra   
346        maret    sultygov         7.67       7.30          0.37   novodra   
347        lixue       hsueh         9.21       8.80          0.41   novodra   
348        jakob    jakobsen         7.96       7.51          0.45   novodra   
349        berta  napolitani         7.68       7.21          0.47   novodra   

    dose_start dose_end           adverse_reaction  
0         41u       48u                        NaN  
1         33u       36u                        NaN  
2         37u       42u                        NaN  
3         31u       38u                        NaN  
4         30u       36u                        NaN  
..         ...      ...                        ...  
345       55u       51u                     nausea  
346       26u       23u                        NaN  
347       22u       23u  injection site discomfort  
348       28u       26u               hypoglycemia  
349       42u       44u  injection site discomfort  

[350 rows x 9 columns]

In [33]:
id_names = patients_copy[['patient_id', 'given_name', 'surname']]
id_names['given_name'] = id_names['given_name'].str.lower()
id_names['surname'] = id_names['surname'].str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

C:\Users\user\AppData\Local\Temp\ipykernel_16440\4157380644.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  id_names['given_name'] = id_names['given_name'].str.lower()
C:\Users\user\AppData\Local\Temp\ipykernel_16440\4157380644.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  id_names['surname'] = id_names['surname'].str.lower()


In [34]:
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0           7.63       7.20          0.43   auralin       41u       48u   
1           7.97       7.62          0.35   auralin       33u       36u   
2           7.65       7.27          0.38   auralin       37u       42u   
3           7.89       7.55          0.34   auralin       31u       38u   
4           7.76       7.37          0.39   auralin       30u       36u   
..           ...        ...           ...       ...        ...      ...   
344         7.51       7.06          0.45   novodra       55u       51u   
345         7.67       7.30          0.37   novodra       26u       23u   
346         9.21       8.80          0.41   novodra       22u       23u   
347         7.96       7.51          0.45   novodra       28u       26u   
348         7.68       7.21          0.47   novodra       42u       44u   

              adverse_reaction  patient_id  
0                          NaN         225  
1                          NaN         242  
2                          NaN         345  
3                          NaN         276  
4                          NaN          15  
..                         ...         ...  
344                     nausea         153  
345                        NaN         420  
346  injection site discomfort         336  
347               hypoglycemia          25  
348  injection site discomfort         477  

[349 rows x 8 columns]

In [35]:
all_columns = pd.Series(list(patients_copy) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

In [37]:
patients_copy.drop_duplicates(subset=['given_name','surname','address'],
                     keep=False, inplace=True)

In [38]:
patients_copy

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country  birthdate  weight  height   bmi       phone_number  \
0    United States  7/10/1976   121.7      66  19.6       951-719-9170   
1    United States   4/3/1967   118.8      66  19.2  +1 (217) 569-3204   
2    United States  2/19/1980   177.8      71  24.8       402-363-6804   
3    United States  7/26/1951   220.9      70  31.7  +1 (732) 636-8246   
4    United States  2/18/1928   192.3      27  26.1       334-515-7487   
..             ...        ...     ...     ...   ...                ...   
498  United States  4/10/1959   181.1      72  24.6       207-477-0579   
499  United States  3/26/1948   239.6      70  34.4       928-284-4492   
500  United States  1/13/1971   171.2      67  26.8       816-223-6007   
501  United States  2/13/1952   176.9      67  27.7       360 443 2060   
502  United States   5/3/1954   138.2      71  19.3       402-848-4923   

                                 email  
0             ZoeWellish@superrito.com  
1                 PamelaSHill@cuvox.de  
2                 JaeMDebord@gustr.com  
3            PhanBaLiem@jourrapide.com  
4                  TimNeudorf@cuvox.de  
..                                 ...  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com  

[497 rows x 15 columns]

In [40]:
patients_copy[patients_copy['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
242         243         male       John     O'Brian                 NaN   
249         250         male   Benjamin      Mehler                 NaN   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country   birthdate  weight  \
29   Port Jervis  New York   12771.0  United States    8/1/1985   155.8   
219          NaN       NaN       NaN            NaN    4/9/1978   237.8   
230          NaN       NaN       NaN            NaN   9/23/1976   165.9   
234          NaN       NaN       NaN            NaN    4/7/1936   199.5   
242          NaN       NaN       NaN            NaN   2/25/1957   205.3   
249          NaN       NaN       NaN            NaN  10/30/1951   146.5   
257          NaN       NaN       NaN            NaN   5/17/1995   231.7   
264          NaN       NaN       NaN            NaN   11/3/1989   158.6   
269          NaN       NaN       NaN            NaN   10/9/1937   175.2   
278          NaN       NaN       NaN            NaN  12/16/1962   124.3   
282     Rainelle        WV   25962.0  United States  10/23/1960   206.1   
286          NaN       NaN       NaN            NaN    4/1/1979   155.3   
296          NaN       NaN       NaN            NaN   5/14/1990   181.1   
502         Burr  Nebraska   68324.0  United States    5/3/1954   138.2   

     height   bmi       phone_number                      email  
29       67  24.4  +1 (845) 858-7707  JakobCJakobsen@einrot.com  
219      69  35.1                NaN                        NaN  
230      63  29.4                NaN                        NaN  
234      65  33.2                NaN                        NaN  
242      74  26.4                NaN                        NaN  
249      69  21.6                NaN                        NaN  
257      69  34.2                NaN                        NaN  
264      63  28.1                NaN                        NaN  
269      61  33.1                NaN                        NaN  
278      69  18.4                NaN                        NaN  
282      64  35.4       304-438-2648   SandraCTaylor@dayrep.com  
286      68  23.6                NaN                        NaN  
296      63  32.1                NaN                        NaN  
502      71  19.3       402-848-4923   PatrickGersten@rhyta.com

In [47]:
patients_copy['zip_code'] = patients_copy['zip_code'].astype(str)
patients_copy['zip_code'] = patients_copy['zip_code'].str[:-2]
patients_copy['zip_code'] = patients_copy['zip_code'].str.zfill(5)
patients_copy.zip_code = patients_copy.zip_code.replace('0000n', np.nan)

In [48]:

patients_copy.zip_code.head()

0    00009
1    00006
2    00006
3    00000
4    00003
Name: zip_code, dtype: object

In [51]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}
def change_name(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
patients_copy['state'] = patients_copy.apply(change_name, axis=1)

In [53]:
patients_copy.state.unique()

array(['CA', 'IL', 'NE', 'NJ', 'AL', 'FL', 'NV', 'MO', 'NY', 'MI', 'TN',
       'VA', 'OK', 'GA', 'MT', 'MA', 'NM', 'LA', 'PA', 'CO', 'ME', 'WI',
       'SD', 'MN', 'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD',
       'AZ', 'TX', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH',
       'OR', nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)